File: haskell/2020/day-18/Rec.hs
import Data.Char

data S = N Integer | S Char
  deriving (Show)
lexer :: String -> [S]
lexer [] = []
lexer (c:cs) | isDigit c = let (a,b) = span isDigit (c:cs) in N (read a) : lexer b
             | isSpace c = lexer (dropWhile isSpace cs)
             | isPunctuation c = S c : lexer cs
             | isSymbol c = S c : lexer cs

data E = C Integer | Op E O E
  deriving (Show)
data O = Add | Mult
  deriving (Show)

parseExpr :: [S] -> (E, [S])
parseExpr ss =
    let (t, ss2) = parseTerm ss
    in case parseExprTail ss2 of
         Nothing -> (t, ss2)
         Just (e,ss3) -> (Op t Mult e, ss3)

parseExprTail (S '*' : ss) = let (t,ss2) = parseExpr ss in Just (t,ss2)
parseExprTail _ = Nothing

parseTerm ss =
    let (u, ss2) = parseUnit ss
    in case parseTermTail ss2 of
         Nothing -> (u, ss2)
         Just (e,ss3) -> (Op u Add e, ss3)

parseTermTail (S '+' : ss) = let (t,ss2) = parseTerm ss in Just (t,ss2)
parseTermTail _ = Nothing

parseUnit (S '(':ss) = let (e, S ')':ss2) = parseExpr ss in (e,ss2)
parseUnit (N x:ss) = (C x, ss)

eval (C x) = x
eval (Op e1 Add e2) = eval e1 + eval e2
eval (Op e1 Mult e2) = eval e1 * eval e2

File: haskell/2020/day-18/Rec.rb
class C
  def initialize(s)
    @l = self.class.lexer(s)
    @p = parseExpr
    @v = eval
    puts(@v)
  end

  def self.lexer(s)
    s.scan(/[0-9]+|[*+()]/)
  end
  def self.ex
    "((2 + 4 * 9) * (6 + 9 * 8 + 6) + 6) + 2 + 4 * 2"
  end

  def parseExpr
    t = parseTerm
    et = parseExprTail
    if et
      ["*", t, et]
    else
      t
    end
  end

  def parseExprTail
    if @l.first == "*"
      @l.shift
      parseExpr
    end
  end

  def parseTerm
    u = parseUnit
    tt = parseTermTail
    if tt
      ["+", u, tt]
    else
      u
    end
  end

  def parseTermTail
    if @l.first == "+"
      @l.shift
      parseTerm
    end
  end

  def parseUnit
    if @l.first == "("
      @l.shift
      e = parseExpr
      n = @l.shift
      raise "!" unless n == ")"
      e
    else
      [@l.shift.to_i]
    end
  end

  def eval
    self.class.eval2(@p)
  end

  def self.eval2(p)
    case p.first
    when "*"
      eval2(p[1]) * eval2(p[2])
    when "+"
      eval2(p[1]) + eval2(p[2])
    else
      p.first
    end
  end
    
end
